### ISSUE TITLE
Guest Data encryption

### ISSUE THREAD

**Issue by cgewecke** [windingtree/wt-js-libs#60]
Prior to this week there was draft work in the repo on PGP encryption of guest data associated with a booking. This seems like a core requirement - we can't let guest data that includes emails, CC, etc, sit in the open on the chain. 

However there's a significant data size issue passing encrypted data into a transaction - even the smallest encrypted packet is several KB in size. We need to handle the vast majority of this data transfer off-chain. One approach:
+ add a short PGP key identifier field to each `Hotel` that can be used to fetch the full public key from a centralized resource like the MIT server, or alternatively, an IPFS hash to the public key. 
+  make `guestData` an IPFS hash to an encrypted packet.
+ provide JS lib support to manage client encryption and manager decryption.

**Comment by ppoliani** [comment]
Hi @cgewecke,

Can you please tell me if my understanding is correct?

So we need to:

1. Encrypt `guestData` off-chain using a PubKey
2. Store the encrypted data to IPFS.
3. Use the above IPFS as the `guestData` that will be stored on-chain.
4. In the `Hotel` contract trigger the following event:

```
Book(from, unitAddress, fromDay, daysAmount, guestData); 
```
5. In an off-chain app, get the `guestData` from the Book event you're interested in and load the encrypted data from IPFS.
6. Decrypt the data using the corresponding private key.

How far is my understanding from your suggestion 😄

**Comment by cgewecke** [comment]
Yes @ppoliani that looks right to me. There are some additional links to resources for managing IPFS hashes within Solidity in the wt-contracts repo [here](https://github.com/windingtree/wt-contracts/issues/148).

**Comment by ppoliani** [comment]
@cgewecke thanks for sharing the links; I will definitely take a look at them.  

Just one more clarification. I am very new to this repository so some of my questions might sound very trivial :)

Basically, the `wt-js-libs` is to be consumed by third-party off-chain apps. For example, a hotel that wants to use the Windingtree platform (to add inventory to the blockchain) will implement an application that will use this set of libraries. In that case, the hotel itself will be responsible for storing the private-public keys that will be used to encrypt/decrypt the **guestData**.

Is that correct? 😅

**Comment by cgewecke** [comment]
@ppoliani I think it's a non-trivial question :) . . . I'm not sure I have a good answer. 

If `wt-js-libs` opts to implement decentralized storage for this kind of data, it seems natural that key generation, encryption/decryption and storage should be integrated as a seamless part of the libs. In turn, perhaps WT could maintain some IPFS nodes on AWS or something and be a 'storage miner' for consumers of this API. 

But it's a big architectural decision. Maybe @kvbik @AugustoL and @adklempner have some views about what to do here. . .

**Comment by kvbik** [comment]
thx @cgewecke for the input and @ppoliani for the questions! i think we did not make the decision yet, because exactly of all those points (where to store/manage all the pubkeys, ...). but we can continue the discussion here!

**Comment by ppoliani** [comment]
I believe that, if the purpose of this library is to help third-parties build their own applications to interact with WindingTree's smart contracts, then we should make their life as easy as possible; So I do agree with @cgewecke that we should provide all the necessary tools to allow them at least to:

1. Generate a key pair that consists of a public and private key (using `gpg` for example)
2. Store/Retrieve keys in/from a key-server (whether it's MIT server or any alternative)
3. Store/Retrieve encrypted data on a decentralized peer-to-peer network (i.e. IPFS).

Most likely the clients of the library will never need to know anything about the above features; the library will internally deal with it as needed.

If I am a developer that wants to publish some inventory to WindingTree, I would love to use this library in the simplest possible way, without having to care about encryption/decryption, IPFS etc.

**Comment by kvbik** [comment]
100% agree, that this library *will* cover this functionality for the developers/users. only because we did not pick the best candidate for the encryption and distribution and storage of the keys, we went with no encrypting during the *test* phase. but for sure it is very important feature, that will land soon!

**Comment by JirkaChadima** [comment]
I'd rephrase *land soon* to *land when necessary* - this means once we are handling some personal data, we will surely introduce some vital support for encryption.
